### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by FredyC**
I think it's time to get this thing out. It seems that React 16.9 has included some preliminary changes for upcoming Concurrent mode which are causing trouble with the current test.

With the code of `useObserver` from `next` branch, tests are passing again although there is a bunch of `act` warnings which might be false positive as it's actually necessary to simulate commit phases and similar. We can review that later.

Personally, we have been using this branch for a couple of months now in production and no related problem seems to occur.

Changelog follows:

```
## Breaking changes:
 * Remove deprecated hooks `useComputed`, `useDisposable`, `useObservable`
 * Internals of `useObserver` are rewritten to support React Strict mode [#121] @RoystonS 
 * UMD bundle is moved to `mobxreactlite.umd.production.min.js` or `mobxreactlite.umd.development.js` respectively.
 * React Native no longer has a specific build, the regular one can be used
```

**This branch has diverted from `master` fair amount and is really hard to rebase. For any further updates just merge master directly and we will squash later.**

### CODE REVIEW

**Comment by xaviergonz** [comment]
LGTM, just I think that there could be an optimization in the cleanup code.
If instead of a plain list the reactions were sorted on insertion (with priority set to cleanup time, via for example via a heap structure), then instead of traversing the whole list you could traverse the list from "soonest to latest to be cleaned" and break as soon as the time is already "not to be cleaned"

**Comment by FredyC** [comment]
@xaviergonz I am sure it can be improved later. Such internal optimization won't be a breaking change most likely. But feel free to submit a PR if you like.

File: src/reactionCleanupTracking.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+function cleanUncommittedReactions() {
+    reactionCleanupHandle = undefined
+
+    // Loop through all the candidate leaked reactions; those older
+    // than CLEANUP_LEAKED_REACTIONS_AFTER_MILLIS get tidied.
+
+    const now = Date.now()
+    for (const ref of uncommittedReactionRefs) {
+        const tracking = ref.current
+        if (tracking) {
```
[the remark this answers is not in the document]
Comment (FredyC) [reply src/reactionCleanupTracking.ts]: I am not really sure. @RoystonS did the whole thing. 

This PR is more about getting stuff together for 2.0. Comments for this particular implementation should probably go to #121 or make a new PR and try to tweak it if tests will pass. Whatever suits you best :)

**Comment by xaviergonz** [comment]
Btw, since it is now using tsdx I have the hunch .browserlistrc, tslint.config and jest config files are no longer needed

File: src/reactionCleanupTracking.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply src/reactionCleanupTracking.ts]: Fair enough :) I'll make a PR

**Comment by xaviergonz** [comment]
And probably it might be worth to add a .eslintrc.js with
```
module.exports = {
    extends: ["react-app", "prettier/@typescript-eslint"]
}
```
if not in the package json already

**Comment by FredyC** [comment]
> Btw, since it is now using tsdx I have the hunch .browserlistrc, tslint.config and jest config files are no longer needed

TSDX is just a build tool, the Jest config is still needed for sure, the `.browserlistrc` not sure how is that utilized actually. TSLint still works, but yea, I want to replace it with ESLint eventually, not sure if straight in 2.0, will see.

**Comment by mweststrate** [comment]
Just checking in, was this waiting for something from my side? (Github notifications stopped working for me years ago)

**Comment by FredyC** [comment]
@mweststrate No, I think this is mostly on me. I got kinda busy and forgot about this. Want to finish up #214 first and now also #226 should be figured out first.

File: src/useObserver.ts [hunk 2]
```diff
@@ -25,39 +35,84 @@ export function useObserver<T>(
...
+                reaction: new Reaction(observerComponentNameFor(baseComponentName), () => {
+                    // We've definitely already been mounted at this point
+                    forceUpdate()
+                }),
+                cleanAt: Infinity
+            }
+            forceUpdate()
+        }
+
+        return () => reactionTrackingRef.current!.reaction.dispose()
```
Comment (normano64) [inline src/useObserver.ts]: This seems to be the line missing for #226
`reactionTrackingRef.current = null;`

Comment (FredyC) [reply src/useObserver.ts]: Hm, was I looking at the wrong line before? 🤔 I suppose I will cut some alpha release with it and try it out. Thanks for spotting it.

**Comment by FredyC** [comment]
Alright, almost ready for release, just waiting for confirmation if the fix for #226 works.

**Comment by mweststrate** [comment]
Given that the next branch solved some problems for others, I guess we can also cut a release and deal with any fallout later? I doubt we get more feedback without releasing.

**Comment by FredyC** [comment]
@mweststrate Well, given the Concurrent is now available for the library authors, I think it would be responsible to test it out properly against it before going full 2.0. However, right now the `@testing-library` is lacking behind and it's not really possible to run tests with Concurrent mode enabled.

**Comment by MohamedLamineAllal** [comment]
The build for the last version `npm install mobx-react-lite@next`. Have a problem considering typescript. index.d.ts is missing! 
The current version have no problem however!

**Comment by FredyC** [comment]
Thanks a lot for the feedback, not sure how could that happen. Fixed in 2.0.0-alpha.4

**Comment by FredyC** [comment]
@auvipy Not before figuring out if there are some issues with the Concurrent mode.

**Pull request merged**